packages/sitecore-jss/src/personalization/layout-personalization-service.ts

```diff
@@ -0,0 +1,220 @@
+ */
+export class LayoutPersonalizationService {
+  private personalizationResult: {
+    personalizationOperation?: Promise<{
+      [key: string]: ComponentRendering | null;
+    }>;
+    components?: { [key: string]: ComponentRendering | null };
+  } = {};
+  private layoutPersonalizationUtils = new LayoutPersonalizationUtils();
+
+  constructor(
+    private personalizationDecisionsService: PersonalizationDecisionsService,
+    private layoutFragmentService: LayoutFragmentService
+  ) {}
+
+  /**
+   * Fetches the personalized data.
+   * @param {PersonalizationContext} context The context.
+   * @param {RouteData} route The route.
+   * @returns {PersonalizationResult} The personalization result.
+   */
+  async fetchPersonalization(
+    context: PersonalizationContext,
+    route: RouteData
+  ): Promise<PersonalizationResult> {
+    // clear personalization before getting new one
+    this.personalizationResult = {};
+
+    const personalizedRenderings = this.layoutPersonalizationUtils.getPersonalizableComponents(
+      route.placeholders
+    );
+
+    if (!personalizedRenderings.length) {
+      return { isPersonalizable: false };
+    }
+
+    const currentResult = this.personalizationResult;
+    currentResult.personalizationOperation = this.personalizeComponents(
+      {
+        routePath: context.routePath,
+        language: context.language,
+      },
+      personalizedRenderings
+    );
+
+    try {
+      const components = await currentResult.personalizationOperation;
+      currentResult.components = components;
+      return { isPersonalizable: true };
+    } catch (error) {
+      currentResult.personalizationOperation = undefined;
+      throw error;
+    }
+  }
+
+  /**
+   * Gets the personalized component.
+   * @param {string} componentUid The unique identifier of a component.
+   * @returns {ComponentRendering} The personalized component.
+   */
+  getPersonalizedComponent(componentUid: string): ComponentRendering | null {
+    if (!this.personalizationResult.components) {
+      return null;
+    }
+
+    return this.personalizationResult.components[componentUid] ?? null;
+  }
+
+  /**
+   * Provides a value that indicates whether the loading is in-progress.
+   * @returns {boolean} The value that indicates whether the loading is in-progress.
+   */
+  isLoading(): boolean {
+    return (
+      !!this.personalizationResult.personalizationOperation &&
+      !this.personalizationResult.components
+    );
+  }
+
+  /**
+   * Ensures the personalized component is loaded.
+   * @param {string} componentUid The unique identifier of a component.
+   * @returns {ComponentRendering} The personalized component.
+   */
+  async ensurePersonalizedComponentLoaded(
+    componentUid: string
+  ): Promise<ComponentRendering | null> {
+    if (!this.personalizationResult.personalizationOperation) {
+      throw new Error(
+        `${this.fetchPersonalization.name} should be called before getting personalized component`
```

Comment:
I'm really uncomfortable with the methods having to be called in a specific order. That seems very error prone.

Reply:
The service is re-designed to make API clear and less error prone